Support default branch on github.com

exoego opened the pull request:
Closes #4

exoego commented:
I have several dependencies that uses non-master branch.
Rate limit wont be my issue, but could be others' issue.

How about using github credential from configuration to extend rate limit?

Clement134 commented:
I agree with you.
Using github credentials seems to be the best way to avoid github rate limit 👍

exoego commented on packages/get-changelog-lib/src/ChangelogFinder.test.js:
```diff
@@ -51,6 +51,36 @@ test('returns CHANGELOG.md url', async () => {
+test('supports default branch', async () => {
+    registryUrl.mockReturnValue('https://registry.npmjs.org/');
+    got.mockImplementation((url) => {
+        console.log(url);
```
Oops. Just removed in https://github.com/Clement134/get-changelog/pull/6/commits/6038f7f3bfeacc2c7e7aa968fb1b83bd0634a0f2

exoego commented on packages/get-changelog-lib/src/ChangelogFinder.js:
```diff
@@ -60,7 +60,19 @@ class ChangelogFinder {
+        const defaultBranch = await (async () =>{
+            try {
+                if (repositoryUrl.includes('github.com')) {
+                    const repoApiPath = repositoryUrl.replace('https://github.com/', 'https://api.github.com/repos/');
```
Fixed in https://github.com/Clement134/get-changelog/pull/6/commits/4c2e3a544a78c0ed0f4dca9660916b97af1bd4bb

exoego commented:
I added naive auth token support in https://github.com/Clement134/get-changelog/pull/6/commits/66b4a9c5c7560836d954d41fdb0d58005d134676
But I had no idea how to test it ...
HDYT?

Clement134 commented:
It's a good idea to use an environment variable to prevent the token from being exposed.

Additionally, I have found why I had reached the rate limit during my tests. Considering that the default branch determination is done in the [_tryChangelogLocation](https://github.com/Clement134/get-changelog/pull/6/files#diff-f63c46ff2442cbde70ed94db8e5200bdR49) function, we could had up to five calls (one for each possible file) to the github API by package. It will be more efficient to compute the default branch in the [getChangelog](https://github.com/Clement134/get-changelog/pull/6/files#diff-f63c46ff2442cbde70ed94db8e5200bdR104) function and add it to the parameters of [_tryChangelogLocation](https://github.com/Clement134/get-changelog/pull/6/files#diff-f63c46ff2442cbde70ed94db8e5200bdR49).

For the tests, you could set the environment variable in the test, check for the Authorization header and delete the environement variable to avoid side effects on other tests.
For example:
```javascript
test('use github token to get default branch', async () => {
    process.env.CHANGELOGFINDER_GITHUB_AUTH_TOKEN = 'githubToken';
    registryUrl.mockReturnValue('https://registry.npmjs.org/');
    const gotMock = got.mockImplementation((url) => {
        if (url === 'https://api.github.com/repos/User/module-name')
            return {
                json: jest.fn().mockResolvedValue({
                    default_branch: 'develop',
                }),
            };
        if (url === 'https://github.com/User/module-name/blob/master/CHANGELOG.md') return 'CHANGELOG DATA';
        return {
            json: jest.fn().mockResolvedValue({
                'dist-tags': {
                    latest: '1.0.0',
                },
                versions: {
                    '1.0.0': {
                        repository: {
                            type: 'git',
                            url: 'git+https://github.com/User/module-name.git',
                        },
                    },
                },
            }),
        };
    });
    const changelogFinder = new ChangelogFinder();
    expect(await changelogFinder.getChangelog('module-name')).toBe('https://github.com/User/module-name/blob/develop/CHANGELOG.md');
    expect(gotMock).nthCalledWith(2, 'https://api.github.com/repos/User/module-name', { headers: { Authorization: 'token githubToken' } });
    delete process.env.CHANGELOGFINDER_GITHUB_AUTH_TOKEN;
});
```

exoego commented:
@Clement134 Thanks for guidance !!
I added that test https://github.com/Clement134/get-changelog/pull/6/commits/85e563c873a0b9d1f197b8ee0054230115f08bc0 and updated code so that branch is calculated once and reused in https://github.com/Clement134/get-changelog/pull/6/commits/5777c4b80be541dae3cd53f4ea7e24cd344a40ff
